projects/observability/src/shared/components/explore-query-editor/order-by/explore-query-order-by-editor.component.ts

```diff
@@ -0,0 +1,239 @@
+
+  private readonly contextSubject: Subject<TraceType | undefined> = new ReplaySubject(1);
+  public readonly selectedMetric$: Observable<AttributeMetadata | undefined>;
+  public readonly selectedAggregation$: Observable<MetricAggregationType | undefined>;
+  public readonly selectedSortBy$: Observable<GraphQlSortDirection | undefined>;
+  public readonly metricOptions$: Observable<SelectOption<AttributeMetadata>[]>;
+  public readonly aggregationOptions$: Observable<SelectOption<MetricAggregationType>[]>;
+  public readonly sortByOptions$: Observable<SelectOption<GraphQlSortDirection | undefined>[]>;
+  private readonly incomingOrderByExpressionSubject: Subject<GraphQlOrderBy | undefined> = new ReplaySubject(1);
+
+  public constructor(private readonly metadataService: MetadataService) {
+    this.selectedAggregation$ = combineLatest([this.contextSubject, this.incomingOrderByExpressionSubject]).pipe(
+      switchMap(() => this.getCurrentSelectedAggregation())
+    );
+
+    this.selectedMetric$ = combineLatest([this.contextSubject, this.incomingOrderByExpressionSubject]).pipe(
+      switchMap(() => this.getCurrentSelectedAttribute())
+    );
+
+    this.selectedSortBy$ = this.incomingOrderByExpressionSubject.pipe(switchMap(() => this.getCurrentSelectedSortBy()));
+
+    this.metricOptions$ = this.contextSubject.pipe(switchMap(context => this.getAttributeOptionsForContext(context)));
+
+    this.aggregationOptions$ = this.selectedMetric$.pipe(
+      map(selection => this.getAggregationOptionsForAttribute(selection))
+    );
+
+    this.sortByOptions$ = this.getSortByOptions();
+  }
+
+  public ngOnChanges(changeObject: TypedSimpleChanges<this>): void {
+    if (changeObject.context) {
+      this.contextSubject.next(this.context);
+    }
+
+    if (changeObject.orderByExpression) {
+      this.incomingOrderByExpressionSubject.next(this.orderByExpression);
+    }
+  }
+
+  public onMetricChange(newAttribute: AttributeMetadata): void {
+    this.buildSpecAndEmit(
+      combineLatest([of(newAttribute), this.selectedAggregation$, of(this.orderByExpression?.direction)])
+    );
+  }
+
+  public onAggregationChange(newAggregation: MetricAggregationType): void {
+    this.buildSpecAndEmit(
+      combineLatest([this.selectedMetric$, of(newAggregation), of(this.orderByExpression?.direction)])
+    );
+  }
+
+  public onSortByChange(newSortBy?: GraphQlSortDirection): void {
+    this.buildSpecAndEmit(combineLatest([this.selectedMetric$, this.selectedAggregation$, of(newSortBy)]));
+  }
+
+  private buildSpecAndEmit(
+    orderByData$: Observable<
+      [AttributeMetadata | undefined, MetricAggregationType | undefined, GraphQlSortDirection | undefined]
+    >
+  ): void {
+    orderByData$
+      .pipe(
+        take(1),
+        switchMap(orderByData => this.buildOrderExpression(...orderByData))
+      )
+      .subscribe(value => this.orderByExpressionChange.emit(value));
```

Comment:
Here it is. So would suggest combining some of the observables going into the template such that we can capture the state in one observable rather than having to manage each one independently and subscribe to all the rest when any one changes.